### PR TITLE
Topt enrollment backend procedure

### PR DESCRIPTION
**Pull request by CDimonaco**
# Description

The pr enable the totp enrollment procedure on the backend.

## How was this tested?

Automated tests

### CODE REVIEW

File: lib/trento_web/openapi/v1/schema/user.ex [hunk 1]
```diff
@@ -6,6 +6,71 @@ defmodule TrentoWeb.OpenApi.V1.Schema.User do
 
   alias TrentoWeb.OpenApi.V1.Schema.Ability.AbilityCollection
 
+  defmodule UserTOTPEnrollmentPayload do
+    @moduledoc false
+
+    @schema %Schema{
+      title: "UserTOTPEnrollmentPayload",
+      description: "Trento User totp enrollment payload",
```
Comment (arbulu89) [inline lib/trento_web/openapi/v1/schema/user.ex:14]: Let's put `topt `as upper case `TOTP` as in other texts

File: lib/trento_web/controllers/v1/profile_controller.ex [hunk 2]
```diff
@@ -41,4 +41,61 @@ defmodule TrentoWeb.V1.ProfileController do
...
+      ok:
+        {"TOTP Enrollment completed", "application/json",
+         Schema.User.UserTOTPEnrollmentConfirmPayload},
+      unprocessable_entity: Schema.UnprocessableEntity.response(),
+      forbidden: Schema.Forbidden.response()
+    ]
+
+  def confirm_totp_enrollment(%{body_params: body_params} = conn, _) do
+    %User{} = user = Pow.Plug.current_user(conn)
+    totp = Map.get(body_params, :totp)
```
Comment (arbulu89) [inline lib/trento_web/controllers/v1/profile_controller.ex:94]: Again, I'm using `totp_code` in the other PR. Let's use the same term.
I will adapt my PR once we have it decided

File: test/trento_web/controllers/v1/profile_controller_test.exs [hunk 3]
```diff
@@ -63,4 +63,183 @@ defmodule TrentoWeb.V1.ProfileControllerTest do
...
+      )
+
+    conn
+    |> put_req_header("content-type", "application/json")
+    |> get("/api/v1/profile/totp_enrollment")
+    |> json_response(:forbidden)
+    |> assert_schema("Forbidden", api_spec)
+  end
+
+  test "should not reset totp when a reset is requested by the default admin", %{
```
Comment (arbulu89) [inline test/trento_web/controllers/v1/profile_controller_test.exs:124]: This shouldn't be `by the default admin`, it should be `for the default admin`

**Review by arbulu89 — APPROVED** [review]
Great @CDimonaco 
I left some comments to try to "unify" some terms with my PR.
Let's discuss here, merge this first so I can adapt the other PR. 
Good job

File: test/trento/discovery/policies/cluster_policy_test.exs [hunk 4]
```diff
@@ -1184,7 +1184,7 @@ defmodule Trento.Discovery.Policies.ClusterPolicyTest do
                |> ClusterPolicy.handle(nil)
     end
 
-    test "should set the health to critical when the SAPInstance resourece is running the same node" do
+    test "should set the health to critical when the SAPInstance resource is running the same node" do
```
Comment (EMaksy) [inline test/trento/discovery/policies/cluster_policy_test.exs:1187]: I guess @CDimonaco found the same issue a second time here @janvhs  :+1: 

https://github.com/trento-project/web/pull/2650/files

File: test/trento_web/controllers/v1/profile_controller_test.exs (hunk 3 above)
Comment (EMaksy) [reply test/trento_web/controllers/v1/profile_controller_test.exs:124]: +1

**Review by EMaksy — COMMENTED** [review]
Hey man great work, +1 to what @arbulu89 said, but besides this lgtm

File: lib/trento/users.ex [hunk 5]
```diff
@@ -135,6 +135,61 @@ defmodule Trento.Users do
...
+      totp_enabled_at: nil,
+      totp_secret: nil,
+      totp_last_used_at: nil
+    })
+  end
+
+  def initiate_totp_enrollment(%User{totp_enabled_at: nil} = user) do
+    result =
+      Ecto.Multi.new()
+      |> Ecto.Multi.run(:user_without_totp, fn _, _ ->
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply lib/trento/users.ex]: done!, I was calling the steps with the user prefix to explicit we are dealing with a use, but the pattern matching will tell us anyway

File: lib/trento/users.ex [hunk 6]
```diff
@@ -135,6 +135,61 @@ defmodule Trento.Users do
...
+      {:error, _, changeset_error, _} ->
+        {:error, changeset_error}
+    end
+  end
+
+  def initiate_totp_enrollment(_), do: {:error, :totp_already_enabled}
+
+  def confirm_totp_enrollment(%User{id: 1}, _), do: {:error, :forbidden}
+
+  def confirm_totp_enrollment(%User{totp_secret: totp_secret, totp_enabled_at: nil} = user, totp) do
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply lib/trento/users.ex]: better totp_code

File: lib/trento/users.ex [hunk 7]
```diff
@@ -198,4 +253,12 @@ defmodule Trento.Users do
   rescue
     Ecto.StaleEntryError -> {:error, :stale_entry}
   end
+
+  defp update_user_totp(%User{id: 1}, _), do: {:error, :forbidden}
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply lib/trento/users.ex:260]: Yeah, private is better we reduce the api surface and it's already tested by the public functions

File: lib/trento_web/openapi/v1/schema/user.ex (hunk 1 above)
Comment (CDimonaco) [reply lib/trento_web/openapi/v1/schema/user.ex]: done!

File: lib/trento_web/controllers/v1/profile_controller.ex (hunk 2 above)
Comment (CDimonaco) [reply lib/trento_web/controllers/v1/profile_controller.ex]: totp_code also here and in the schemas!

File: test/trento/discovery/policies/cluster_policy_test.exs (hunk 4 above)
Comment (CDimonaco) [reply test/trento/discovery/policies/cluster_policy_test.exs:1187]: :eyes: great ideas are in the air

File: lib/trento_web/controllers/v1/profile_controller.ex [hunk 8]
```diff
@@ -41,4 +41,61 @@ defmodule TrentoWeb.V1.ProfileController do
       render(conn, "profile.json", user: updated_user)
     end
   end
+
+  operation :reset_totp,
+    summary: "Reset the TOTP configuration for the user",
+    tags: ["Platform"],
+    responses: [
+      forbidden: Schema.Forbidden.response()
```
[the remark this answers is not in the document]
Comment (CDimonaco) [reply lib/trento_web/controllers/v1/profile_controller.ex]: done!

**Review by arbulu89 — APPROVED** [review]
Green light from my side!
Thank you for the changes.
I will rebase yours and do some ticks in my code

**Pull request merged**